WordPress/Classes/Networking/MediaRequestAuthenticator.swift

```diff
@@ -0,0 +1,247 @@
+import Foundation
+
+fileprivate let photonHost = "i0.wp.com"
+
+extension URL {
+    /// Whether the URL is hosted at WPCom.
+    ///
+    func isHostedAtWPCom() -> Bool {
+        return host?.hasSuffix(".wordpress.com") ?? false
+    }
+
+    /// Whether the URL is a Photon URL.
+    ///
+    fileprivate func isPhoton() -> Bool {
+        return host == photonHost
+    }
+}
+
+/// This class takes care of resolving any authentication necessary before
+/// requesting media from WP sites (both self-hosted and WP.com).
+///
+/// This also includes regular and photon URLs.
+///
+class MediaRequestAuthenticator {
+
+    /// Errors conditions that this class can find.
+    ///
+    enum Error: Swift.Error {
+        case cannotFindSiteIDForSiteAvailableThroughWPCom(blog: Blog)
+        case cannotBreakDownURLIntoComponents(url: URL)
+        case cannotCreateAtomicURL(components: URLComponents)
+        case cannotCreateAtomicProxyURL(components: URLComponents)
+        case cannotCreatePrivateURL(components: URLComponents)
+        case cannotFindWPContentInPhotonPath(components: URLComponents)
+        case failedToLoadAtomicAuthenticationCookies(underlyingError: Swift.Error)
+    }
+
+    // MARK: - Request Authentication
+
+    /// Pass this method a media URL and host information, and it will handle all the necessary
+    /// logic to provide the caller with an authenticated request through the completion closure.
+    ///
+    /// - Parameters:
+    ///     - url: the url for the media.
+    ///     - host: the `MediaHost` for the requested Media.  This is used for authenticating the requests.
+    ///     - provide: the closure that will be called once authentication is sorted out by this class.
+    ///         The request can be executed directly without having to do anything else in terms of
+    ///         authentication.
+    ///     - fail: the closure that will be called upon finding an error condition.
+    ///
+    func authenticatedRequest(
+        for url: URL,
+        from host: MediaHost,
+        onComplete provide: @escaping (URLRequest) -> (),
+        onFailure fail: @escaping (Error) -> ()) {
+
+        // We want to make sure we're never sending credentials
+        // to a URL that's not safe.
+        guard url.isHostedAtWPCom() || url.isPhoton() else {
+            let request = URLRequest(url: url)
+            provide(request)
+            return
+        }
+
+        switch host {
+        case .publicSite: fallthrough
+        case .publicWPComSite: fallthrough
```

Comment:
Nipticky: for readability purposes, maybe `fallthrough` in a newline?

Reply:
Done!